peps/pep-0763.rst

```diff
@@ -0,0 +1,275 @@
+
+The size and interdependency of the Python ecosystem is continuing to grow,
+meaning that future deletions of projects can be reasonably assumed to
+be *just as if not more* disruptive than the deletions sampled above.
+
+Given the above, this PEP concludes that the availability of "hard" deletions
+versus "soft" deletions (i.e. yanking) now presents more of a risk and detriment
+to the Python ecosystem than a benefit.
+
+Specification
+=============
+
+This PEP identifies 3 different types of deletable objects:
+
+1. **Files**, which are individual project distributions (such as source
+   distributions or wheels).
+
+   Example: ``requests-2.32.3-py3-none-any.whl``.
+
+2. **Versions**, which contain one or more files that share the same version.
+
+   Example: `requests v2.32.3 <https://pypi.org/project/requests/2.32.3/>`_.
+
+3. **Projects**, which contain one or more versions.
+
+   Example: `requests <https://pypi.org/project/requests>`_.
+
+This PEP proposes the following *deletion eligibility rules*:
+
+* A **file** is considered deletable if and only if it was uploaded to
+  PyPI less than 72 hours from the current time, **or** if it
+  has a :ref:`pre-release specifier <packaging:version-specifiers>`.
+* A **release** is considered deletable if and only if all of its
+  constituent files are deletable.
+* A **project** is considered deletable if and only if all of its
+  constituent releases are deletable.
+
+These rules are intentionally "telescoping": they allow new projects to be
+deleted entirely, and allow old projects to delete new files or releases,
+but do not allow old projects to delete old files or releases.
+
+This is intended to strike a balance between competing interests: brand new
+projects are unlikely to have significant community uptake and thus pose a
+minimal disruptive risk, while established projects (of any size)
+are more likely to have a "tail" of adopted versions. Their downstream users
+are not necessarily equipped to address the sudden deletion
+of a version, file, or the whole project.
+
+Implementation
+==============
+
+This PEP's implementation primarily concerns aspects of PyPI that are not
+standardized or subject to standardization, such as the web interface and
+signed-in user operations. As a result, this section describes its
+implementation in behavioral terms.
+
+Changes
+-------
+
+* Per the eligibility rules above, PyPI will reject web interface requests
+  (using an appropriate HTTP response code of its choosing) for
+  file, release, or project deletion if the respective object is not
+  eligible for deletion.
+* PyPI will amend its web interface to indicate a file/release/project's
+  deletion ineligibility, e.g. by styling the relevant UI elements as "inactive"
+  and making relevant bottoms/forms unclickable.
+
+Security Implications
+=====================
+
+This PEP does not identify any positive or negative security implications
+associated with proposed approach.
```

Comment:
Given above we discuss security implications of *permitting* deletions, I would expect that the security implications are *this PEP makes things better* (but also see my comment above).

Reply:
A possible negative implication is that it may be harder to stop users from downloading a file that has security issues, right?

Reply:
Thanks! Yeah, I need to qualify this: the positive implication above was that external users of the index (e.g. companies) benefit from limiting deletion because it makes post-compromise triage and remediation simpler, since the attacker can't remove their trail as easily. I wasn't sure whether to put that under this section though, since it's an indirect/fuzzy security implication, but I'll work it in for further discussion.

> A possible negative implication is that it may be harder to stop users from downloading a file that has security issues, right?

@JelleZijlstra Could you elaborate a bit on what you're thinking here? This PEP doesn't stipulate any changes to PyPI admins' abilities to delete things, so the existing malware reporting and quarantining flows remain in place. So from my perspective this doesn't make it any harder to halt downloads of malicious packages.

OTOH, maybe you're thinking of packages that aren't malicious but just have security issues? For those I think the pre-existing "yank" mechanism is good enough (and is what people already use in practice), but I could add in some qualifications there if you think the PEP would benefit from it!

Reply:
Currently, if a package maintainer discovers that there is malware in a package they released, they can delete the package. With this PEP's proposal, they'd have to instead wait for PyPI admins to get to it, which may take longer.

Example scenario: I am a maintainer of https://pypi.org/project/black/. I discover (e.g., through private disclosure) that one of the compiled wheels of the latest release contains malicious code due to a backdoor in cibuildwheel/GH Actions/mypyc (pretty outlandish scenarios, but bear with me). Right now, the first thing I would do is go to PyPI and delete the malicious wheel. With this PEP's proposal, I'd have to figure out the right channels to raise the issue to the PyPI admins, and possibly wait for some time.

This is not a common scenario and your answer could well be that the tradeoff is worth it, but it doesn't seem impossible.

Reply:
Gotcha, thanks for elaborating! That's an interesting case -- I can see the argument that it represents a slightly worse security posture, since there's now a middleman (PyPI's admins) for malware response.

I think my position is that it's "worth it," especially now that PyPI has a dedicated malware reporting flow (see screencap) as well as project "quarantine" support for temporarily suspending a project's downloads without actually deleting it. But I'll make sure to include that in a revision to fairly characterize this PEP's changes.

<img width="361" alt="Screenshot 2024-10-28 at 12 10 53 PM" src="https://github.com/user-attachments/assets/8114b805-34fb-42f1-a617-0db463902a57">

Reply:
Thanks, that position makes sense to me!